### PR TITLE
win32unix: use WSADuplicateSocket in bindings of dup and dup2

### PR DESCRIPTION
**Pull request by MisterDA**
The WinAPI documentation of [DuplicateHandle][] gives a list of types
of objects on which it is safe to call `DuplicateHandle`. In particular,
the doc reads:

> You should not use `DuplicateHandle` to duplicate handles to the
> following objects:
>
> - Sockets. No error is returned, but the duplicate handle may not be
>   recognized by Winsock at the target process. Also, using
>   `DuplicateHandle` interferes with internal reference counting on the
>   underlying object. To duplicate a socket handle, use the
>   `WSADuplicateSocket` function.

Rework the code to use `DuplicateHandle` for handles and
[WSADuplicateSocket][] for sockets.

`WSADuplicateSocket` returns an info structure that we give to `WSASocket`
to create the duplicated handle. However, there seem to be no way of
explicitely preserving the set of flags to give to `WSASocket`.
Considering that as of now, OCaml uses the socket function instead of
`WSASocket`, and that socket only sets `WSA_FLAG_OVERLAPPED`, re-set that
flag and hope for the best (that's also the current behavior of
libuv).

It's also possible to retreive the `WSA_FLAG_NO_HANDLE_INHERIT` (Winsock
mapping of `HANDLE_FLAG_INHERIT`) via `GetHandleInformation`, but that's
not needed here since we override it with the `cloexec` parameter.

This commit also fixes a little bug. In the error path of
`DuplicateHandle` there was:

    win32_maperr(GetLastError());
    return -1;

but the function returns a value, and for the OCaml code above it's
better to raise an exception instead.

[DuplicateHandle]: https://docs.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-duplicatehandle
[WSADuplicateSocket]: https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsaduplicatesocketw

### CODE REVIEW

**Comment by xavierleroy** [comment]
This looks pretty good to me.  Thanks for the explanations.

Would it be possible to share more code between dup and dup2 ?  The part that duplicates the handle or the socket looks identical in both files (if I'm not mistaken).  If you wish, you could even move the dup2 implementation to otherlibs/win32unix/dup.c and delete otherlibs/win32unix/dup2.c.

**Comment by xavierleroy** [comment]
Also, we'll need another opinion, e.g. @dra27 's or @nojb 's.

File: otherlibs/win32unix/dup.c
Comment (nojb) [inline otherlibs/win32unix/dup.c]:
```suggestion
static HANDLE duplicate_handle(BOOL inherit, HANDLE oldh)
```
This argument is the negation of `cloexec` and should be named accordingly.

**Review by nojb — APPROVED** [review]
This looks like a nice cleanup, thanks! LGTM, modulo a tiny naming correction.

**Comment by MisterDA** [comment]
Thank you for the review.

> Would it be possible to share more code between dup and dup2?

Sure, that's quite sensible. I think I've managed to share the code in a nice way.

> This argument is the negation of `cloexec` and should be named accordingly.

Yes, I noticed that when I also noticed that I wasn't using the `cloexec` parameter with sockets! That's fixed now.
There's the flag `WSA_FLAG_NO_HANDLE_INHERIT` of `WSASocket` that we could use instead of calling `win_set_inherit` (it has the same effect), but the flag is available with Windows 7 and later only.

File: Changes
Comment (dra27) [inline Changes]:
```suggestion
- #10697: Bindings of dup and dup2 in win32unix now correctly call
  WSADuplicateSocket on sockets instead of DuplicateHandle.
  (Antonin Décimo, review by Xavier Leroy and Nicolás Ojeda Bär)
```
(I think it's worth noting that what was there before was potentially unstable)

**Review by dra27 — COMMENTED** [review]
I'm struggling a bit at the moment, so I'm afraid this is only partial - it looks clear that the bug is real but not that the fix is correct (but I'm not certain). While it's clear that we shouldn't be calling `DuplicateHandle` on a socket, for handles within the same process, we could just make it a no-op (and just do the CRT dup/dup2, if applicable). The structure returned by `WSADuplicateSocket` is supposed to be passed to `WSASocket` _by the target process_ - when it's the same process as created it, it's not clear that that's doing anything (in fact, MSDN suggests its unnecessary).

Does the existing code demonstrably allow a child process to inherit a socket handle (e.g. by receiving it on stdin/stdout) and, if not, would it be better to avoid the duplication (MSDN is quite explicit that process threads can share socket handles)?

**Comment by xavierleroy** [comment]
> While it's clear that we shouldn't be calling DuplicateHandle on a socket, for handles within the same process, we could just make it a no-op (and just do the CRT dup/dup2, if applicable). 

This would be completely inadequate.  E.g. after a `dup`, it should be possible to close one of the two file descriptors (argument and result of `dup`) independently of the other.

I've always be amazed at how Windows handles generalize Unix file descriptors while working very much like file descriptors, except socket handles, which seem to differ from other handles in pretty much every respect.  What's the point??

**Comment by MisterDA** [comment]
Rebased the PR with the suggested changes. Xavier's right about closing fd independently.
It's not very useful to duplicate sockets in the same process, but we obviously cannot retro-fit Windows model here (duplicate, get the info structure, send the structure to the other process, call `WSASocket` with the info) with a single call. I think the PR is as correct as it can get, let me know if it needs further work.

**Comment by dra27** [comment]
Thanks!

**Pull request merged**
